runtime/vm/ContinuationHelpers.cpp

```diff
@@ -141,19 +176,19 @@ yieldContinuation(J9VMThread *currentThread)
 {
 	BOOLEAN result = TRUE;
 	J9VMContinuation *continuation = currentThread->currentContinuation;
-
 	Assert_VM_notNull(currentThread->currentContinuation);
 
-	VM_ContinuationHelpers::swapFieldsWithContinuation(currentThread, continuation);
-	continuation->carrierThread = NULL;
 	currentThread->currentContinuation = NULL;
+	VM_ContinuationHelpers::swapFieldsWithContinuation(currentThread, continuation);
 
 	/* We need a full fence here to preserve happens-before relationship on PPC and other weakly
 	 * ordered architectures since learning/reservation is turned on by default. Since we have the
 	 * global pin lock counters we only need to need to address yield points, as thats the
 	 * only time a different virtualThread can run on the underlying j9vmthread.
 	 */
 	VM_AtomicSupport::readWriteBarrier();
+	Assert_VM_true(currentThread == VM_VMHelpers::getCarrierThreadFromContinuationState(continuation->state));
```

Comment:
You can also add a comment that we don't need atomic here, since no GC thread should be able to start scanning while continuation is mounted, nor should another carrier thread be able to mount before we complete the unmount (hence no risk to overwrite anything in a race)

After that comment you can add an assert that it indeed concurrentlyScanned is not set (and have the other assert about carrier ID). Or that could just really be one assert that the state is exactly carrier ID